datafusion/expr-common/src/scalar.rs

```diff
@@ -0,0 +1,98 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::{
+    array::{Array, ArrayRef},
+    datatypes::DataType,
+};
+use datafusion_common::{exec_err, Result, ScalarValue};
+
+#[derive(Clone, Debug)]
+pub struct Scalar {
+    value: ScalarValue,
+    data_type: DataType,
+}
+
+impl From<ScalarValue> for Scalar {
+    fn from(value: ScalarValue) -> Self {
+        Self {
+            data_type: value.data_type(),
+            value,
+        }
+    }
+}
+
+impl PartialEq for Scalar {
+    fn eq(&self, other: &Self) -> bool {
+        self.value.eq(&other.value)
+    }
+}
+
+impl Scalar {
+    pub fn try_from_array(array: &dyn Array, index: usize) -> Result<Self> {
+        let data_type = array.data_type().clone();
+        let value = ScalarValue::try_from_array(array, index)?;
+        Ok(Self { value, data_type })
+    }
+
+    pub fn new_null_of(value: DataType) -> Result<Self> {
+        Ok(Self {
+            value: ScalarValue::try_from(&value)?,
+            data_type: value,
+        })
+    }
+
+    pub fn iter_to_array(scalars: impl IntoIterator<Item = Scalar>) -> Result<ArrayRef> {
+        let mut scalars = scalars.into_iter().peekable();
+
+        // figure out the type based on the first element
+        let data_type = match scalars.peek() {
+            None => return exec_err!("Empty iterator passed to Scalar::iter_to_array"),
+            Some(sv) => sv.data_type().clone(),
+        };
+
+        ScalarValue::iter_to_array_of_type(scalars.map(|scalar| scalar.value), &data_type)
+    }
+
+    #[inline]
+    pub fn value(&self) -> &ScalarValue {
```

Comment:
We could potentially call this function `as_value` to hit that it was returning a reference and was cheap

Reply:
eg arrow `Field::name()` is named just "name". isn't `as_foo` more appropriate for representhing self as foo, rather than reading a foo attribute of self?